File: server-side/src/RealTime/RealTime.Chat/RealTime.Chat/Handlers/SendMessageCommandHandler.cs
using Amazon;
using Amazon.ApiGatewayManagementApi;
using Amazon.ApiGatewayManagementApi.Model;
using Amazon.Lambda.APIGatewayEvents;
using Amazon.Lambda.Core;
using Amazon.Runtime;
using Common.Layer.JsonOptions;
using RealTime.Chat.Models;
using RealTime.Chat.Persistence;
using System.Drawing;
using System.IdentityModel.Tokens.Jwt;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace RealTime.Chat.Handlers;

public class SendMessageCommandHandler
{
    private readonly IChatRepository _chatRepository;
    private readonly IMessageRepository _messageRepository;
    private readonly IConnectionMappingRepository _connectionMappingRepository;
    private readonly IAmazonApiGatewayManagementApi _apiGatewayManagementApi;
    private ILambdaContext _context;

    public SendMessageCommandHandler()
    {
        _chatRepository = new ChatRepository();
        _messageRepository = new MessageRepository();
        _connectionMappingRepository = new ConnectionMappingRepository();
        _apiGatewayManagementApi = new AmazonApiGatewayManagementApiClient(new AmazonApiGatewayManagementApiConfig()
        {
            ServiceURL = Environment.GetEnvironmentVariable("SERVICE_URL")
        });
    }

    public async Task<APIGatewayProxyResponse> FunctionHandler(APIGatewayProxyRequest request, ILambdaContext context)
    {
        _context = context;
        var connectionId = request.RequestContext.ConnectionId;

        try
        {
            var command = JsonSerializer.Deserialize<SendMessageCommand>(request.Body, JsonOptions.Options);
            var chat = await _chatRepository.GetByIdAsync(command.ChatId);

            var senderConnectionMapping = await _connectionMappingRepository.GetByConnectionAsync(connectionId);

            var message = new Message(chat.Id, senderConnectionMapping.Sub, command.Text);
            await _messageRepository.SaveAsync(message);

            var sendToId = (chat.ClientId == senderConnectionMapping.Sub) ? chat.FreelancerId : chat.ClientId;

            var connectionMapping = await _connectionMappingRepository.GetAsync(sendToId);

            if (connectionMapping == null)
            {
                return new APIGatewayProxyResponse
                {
                    StatusCode = 404,
                    Body = "Connection not found"
                };
            }

            await SendMessageToConnectionAsync(connectionMapping.ConnectionId, command.Text);

            _context.Logger.LogInformation("GOTOVO");

            return new APIGatewayProxyResponse
            {
                StatusCode = 200,
                Body = "Message sent successfully"
            };

        }
        catch (Exception ex)
        {
            context.Logger.LogLine($"Error: {ex}");

            return new APIGatewayProxyResponse
            {
                StatusCode = 500,
                Body = "Internal Server Error"
            };
        }
    }

    private async Task SendMessageToConnectionAsync(string connectionId, string message)
    {
        try
        {
            var messageBytes = System.Text.Encoding.UTF8.GetBytes(message);
            using var memoryStream = new MemoryStream(messageBytes);

            _context.Logger.LogInformation($"CID {connectionId}");

            var request = new PostToConnectionRequest
            {
                ConnectionId = connectionId,
                Data = new MemoryStream()
            };

            await _apiGatewayManagementApi.PostToConnectionAsync(request);
        }
        catch (Exception ex)
        {
            _context.Logger.LogError($"ERROR: {ex}");
            throw new Exception($"Error sending message to connection {connectionId}: {ex.Message}");
        }
    }
}

public class SendMessageCommand
{
    public Guid ChatId { get; set; }
    public string Text { get; set; }

}

File: server-side/src/RealTime/RealTime.Chat/RealTime.Chat/Handlers/SendMessageCommandHandler.js
import { v4 as uuidv4 } from 'uuid';
import AWS from 'aws-sdk';

const dynamoDB = new AWS.DynamoDB({ region: 'eu-central-1' });

const apigatewaymanagementapi = new AWS.ApiGatewayManagementApi({
    region: 'eu-central-1',
    endpoint: process.env.SERVICE_URL
});

export const handler = async (event) => {
    const connectionId = event.requestContext.connectionId;
    try {
        const command = JSON.parse(event.body);
        const chat = await getChatByIdAsync(command.chatId);
        console.log(chat);

        const senderConnectionMapping = await getCMByConnectionAsync(connectionId);
        console.log(senderConnectionMapping);

        const message = {
            id: uuidv4().toString(),
            chatId: chat.Id,
            senderId: senderConnectionMapping.Sub,
            text: command.text,
            date: new Date().toISOString()
        };
        await saveMessageAsync(message);

        const sendToId = chat.ClientId === senderConnectionMapping.Sub ? chat.FreelancerId : chat.ClientId;

        const connectionMapping = await getCMBySubAsync(sendToId);

		const responseMessage = {
            action: 'newMessage',
            body: { ...message }
        }

        if (!connectionMapping.ConnectionId) {
            return {
                statusCode: 200,
                body: JSON.stringify(responseMessage)
            };
        }

        await sendMessageToConnectionAsync(connectionMapping.ConnectionId, responseMessage);

        console.log('Message sent successfully');

        return {
            statusCode: 200,
            body: JSON.stringify(responseMessage)
        };

    } catch (error) {
        console.log(`Error: ${error}`);

        return {
            statusCode: 500,
            body: JSON.stringify({ message: 'Internal Server Error' })
        };
    }
};

async function sendMessageToConnectionAsync(connectionId, message) {
    const params = {
        ConnectionId: connectionId,
        Data: JSON.stringify(message)
    };

    try {
        await apigatewaymanagementapi.postToConnection(params).promise();
    } catch (error) {
        console.error(`Error sending message to connection ${connectionId}:`, error);
        throw new Error(`Error sending message to connection ${connectionId}: ${error.message}`);
    }
}

async function getChatByIdAsync(id) {
    console.log(id);
    const params = {
        TableName: 'Chats',
        Key: {
            'Id': { S: id }
        }
    };

    try {
        const result = await dynamoDB.getItem(params).promise();
        return AWS.DynamoDB.Converter.unmarshall(result.Item);
    } catch (error) {
        console.error('Error getting chat by id:', error);
        throw error;
    }
}

async function getCMBySubAsync(sub) {
    const params = {
        TableName: 'ChatConnectionMapping',
        Key: {
            'Sub': {
                S: sub
            }
        }
    };

    try {
        const result = await dynamoDB.getItem(params).promise();
        return AWS.DynamoDB.Converter.unmarshall(result.Item);
    } catch (error) {
        console.error('Error getting connection mapping by sub:', error);
        throw error;
    }
}

async function getCMByConnectionAsync(connectionId) {
    const params = {
        TableName: 'ChatConnectionMapping',
        FilterExpression: 'ConnectionId = :connectionId',
        ExpressionAttributeValues: {
            ':connectionId': { S: connectionId }
        }
    };

    try {
        const result = await dynamoDB.scan(params).promise();
        return AWS.DynamoDB.Converter.unmarshall(result.Items[0]);
    } catch (error) {
        console.error('Error getting connection mapping by connectionId:', error);
        throw error;
    }
}

async function saveMessageAsync(message) {
    const params = {
        TableName: 'Messages',
        Item: {
            'Id': { S: message.id },
            'ChatId': { S: message.chatId },
            'SenderId': { S: message.senderId },
            'Text': { S: message.text },
            'Date': { S: message.date },
        }
    };

    try {
        await dynamoDB.putItem(params).promise();
    } catch (error) {
        console.error('Error saving message:', error);
        throw error;
    }
}
